Use traits for (un)directed types

filchristou opened the pull request:
I find the definition of `MetaUndirectedGraph` and `MetaDiGraph` very constrained.
`Graphs.jl` solves this problem by using traits. 
Here, I made the changes so that the same solution is followed. 

`SimpleTraits.jl` is already in the Manifest as an indirect dependency due to Graphs.jl 

With this solution we can now build `MetaGraph`s of any subtype of `AbstractGraph` implementing `Graphs.is_directed()`
Before only `MetaGraph`s of `SimpleGraph` and `SimpleDiGraph` could be built.

### Use case
For example, if this PR, together with https://github.com/QuantumBFS/Multigraphs.jl/pull/18 and https://github.com/QuantumBFS/Multigraphs.jl/pull/17 are completed, we will be able to have a `MetaGraph` of `Multigraph`.
This will be a huge advantage over [MetaGraphs.jl](https://github.com/JuliaGraphs/MetaGraphs.jl), which for now it's also only limited to `SimpleGraph` and `SimpleDiGraph`.

#### More discussion
I don't quite agree with the `Graphs.SimpleGraph` and `Graphs.SimpleDiGraph` methods.
To me they look like constructors and they should return a new object, and not the graph of the `MetaGraph`.
For this reason, we could instead have something simple like a `getgraph` method.

bramtayl commented:
Thanks! Using traits for this makes sense to me!

bramtayl commented:
I also agree with you about the `Graphs.SimpleGraph` and `Graphs.SimpleDiGraph` methods being confusing. We inherited this from the original `MetaGraphs`

bramtayl commented:
If you want to help maintain I could add you as a collaborator?

filchristou commented:
Do you have any comments on the PR ? Keep in mind that this is a breaking change, since I removed `MetaDiGraph` and `MetaUndirectedGraph`. No particular reason why I removed those and one could add them back, but they go opposite to the idea of `MetaGraphNext` being a general wrapper.

> I also agree with you about the Graphs.SimpleGraph and Graphs.SimpleDiGraph met

I made an issue if more conversation needs to happen https://github.com/JuliaGraphs/MetaGraphsNext.jl/issues/42
This PR already took the liberty to remove these. Let me know if that's fine.

> If you want to help maintain I could add you as a collaborator?

Yes sure! I cannot promise working consistently but I will be of help, so please go on!

bramtayl commented:
What do you think @gdalle?

gdalle commented on test/runtests.jl:
```diff
@@ -2,11 +2,15 @@ using Documenter
     doctest(MetaGraphsNext)
 
     colors = MetaGraph( Graph(), VertexData = String, EdgeData = Symbol, graph_data = "graph_of_colors")
+    @test istrait(IsDirected{typeof(colors)}) == is_directed(colors) == false
```
I put all this in a separate test file

gdalle commented:
Thanks for the great PR @filchristou, it's a really awesome idea!
